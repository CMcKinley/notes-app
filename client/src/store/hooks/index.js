export { default as useGetNotes } from './useGetNotes';
export { default as useEditNotes } from './useEditNotes';
export { default as useAddNotes } from './useAddNotes';