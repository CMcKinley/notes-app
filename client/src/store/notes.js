import { handleActions, createAction } from "redux-actions";
import cloneDeep from "lodash.clonedeep";

const defultState = {
  noteIds: [],
  noteMap: {},
  noteToEdit: null
};

const prefix = "NOTES/";

// Action Creators
export const setNotes = createAction(`${prefix}SET_NOTES`);
export const editNoteInStore = createAction(`${prefix}EDIT_NOTE`);
export const addNoteToStore = createAction(`${prefix}ADD_NOTE`);
export const setNoteToEdit = createAction(`${prefix}SET_NOTE_TO_EDIT`);

// Reducer
export default handleActions(
  {
    // Load notes into state. Notes are stored as an array of ids and an object where each key is the note's id and the value is the note object itself.
    [setNotes]: (state, { payload }) => {
      const notes = payload || [];
      const noteIds = [];
      const noteMap = {};

      notes.forEach(note => {
        const { id } = note || {};
        noteIds.push(id);
        noteMap[id] = note;
      });

      return { ...state, noteIds, noteMap };
    },

    // Edit an existing note
    [editNoteInStore]: (state, { payload }) => {
      const { id } = payload || {};
      // The values of the note map are objects so we need to clone deep the noteMap to ensure each object's properties are not references to the original
      const newMap = cloneDeep(state.noteMap);
      newMap[id] = { ...newMap[id], ...payload };
      return { ...state, noteMap: newMap };
    },

    // Add a new note to the id array and note map
    [addNoteToStore]: (state, { payload }) => {
      const { id } = payload || {};
      const noteIds = [...state.noteIds];
      // The values of the note map are objects so we need to clone deep the noteMap to ensure each object's properties are not references to the original
      const noteMap = cloneDeep(state.noteMap);
      // Add note to beginning of array
      noteIds.unshift(payload.id);
      noteMap[id] = payload;
      return { ...state, noteIds, noteMap };
    },

    // Set the note that is to be edited in store
    [setNoteToEdit]: (state, { payload }) => ({ ...state, noteToEdit: payload })
  },
  defultState
);
