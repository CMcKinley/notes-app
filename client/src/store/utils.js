import axios from "axios";

// Store tools will be used to make http calls to the node api
export default class StoreTools {
  async base(url, body, method = "get", headers = {}) {
    let config = { url: `/api/v1/${url}`, method, headers };
    if (body) {
      config.data = body;
    }
    return axios(config);
  }
  // GET requests
  async get(url, headers = {}) {
    return this.base(url, null, "get", headers);
  }
  // PUT requests
  put(url, body, headers = {}) {
    return this.base(url, body, "put", headers);
  }
  // POST requests
  post(url, body, headers = {}) {
    return this.base(url, body, "post", headers);
  }
}
