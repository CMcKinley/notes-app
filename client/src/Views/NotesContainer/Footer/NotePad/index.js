import React, { memo, useState, useEffect } from "react";
import TextField from "@material-ui/core/TextField";
import { makeStyles } from "@material-ui/core/styles";
import Fab from "@material-ui/core/Fab";
import ClearIcon from "@material-ui/icons/Clear";
import SubmitIcon from "@material-ui/icons/Send";
import { useAddNotes, useEditNotes } from "store/hooks";
import { useSelector } from "react-redux";

const useStyles = makeStyles(() => ({
  errorText: {
    fontSize: ".7em",
    color: "red"
  },
  inputWrapper: {
    flex: 1,
    maxWidth: "600px",
    background: "white",
    borderRadius: "6px",
    padding: "4px 16px 8px",
    margin: "8px"
  },
  textField: {
    background: "white",
    borderRadius: "6px"
  },
  fab: {
    margin: "8px"
  },
  actionButtons: {
    display: "flex",
    flexDirection: "column"
  }
}));

const NotePad = ({ setOpen, noteIdToBeEdited }) => {
  const classes = useStyles();
  const [text, setText] = useState("");
  const [nonprintable, setNonprintable] = useState(false);
  const [addNotes, isLoading] = useAddNotes();
  const [editNotes, editIsLoading] = useEditNotes();
  const noteToBeEdited = useSelector(state =>
    noteIdToBeEdited ? state.notes.noteMap[noteIdToBeEdited] : null
  );

  useEffect(() => {
    if (noteToBeEdited) setText(noteToBeEdited.text);
  }, [noteToBeEdited]);

  // Handle input change. If there are non printable characters set error in state.
  function handleChange(e) {
    const value = e.target.value;
    if (value && !/^[ -~\t\n\r]+$/.test(value)) {
      setNonprintable(true);
    } else if (nonprintable) {
      setNonprintable(false);
    }
    setText(e.target.value);
  }

  async function handleSubmit() {
    if (noteIdToBeEdited) {
      await editNotes({ ...noteToBeEdited, text });
    } else {
      await addNotes({ text });
    }

    setOpen(false);
  }

  return (
    <>
      <div className={classes.inputWrapper}>
        <TextField
          id="note"
          className={classes.textField}
          label="New Note"
          value={text}
          multiline
          fullWidth
          error={nonprintable}
          margin="normal"
          onChange={handleChange}
        />
        {nonprintable && (
          <div className={classes.errorText}>
            Please remove non-printable characters
          </div>
        )}
      </div>
      <div className={classes.actionButtons}>
        <Fab
          size="small"
          aria-label="add"
          onClick={handleSubmit}
          className={classes.fab}
          disabled={nonprintable || !text || isLoading || editIsLoading}
        >
          <SubmitIcon />
        </Fab>
        <Fab
          size="small"
          aria-label="add"
          onClick={() => setOpen(false)}
          className={classes.fab}
        >
          <ClearIcon />
        </Fab>
      </div>
    </>
  );
};

export default memo(NotePad);
