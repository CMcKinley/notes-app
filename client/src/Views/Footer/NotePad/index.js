import React, { memo, useState, useEffect } from "react";
import PropTypes from "prop-types";
import TextField from "@material-ui/core/TextField";
import { makeStyles } from "@material-ui/core/styles";
import { useAddNotes, useEditNotes } from "store/hooks";
import { useSelector } from "react-redux";
import ActionButtons from "./ActionButtons";
import { useDispatch } from "react-redux";
import { setNoteToEdit } from "store/notes";

const useStyles = makeStyles(() => ({
  errorText: {
    fontSize: ".7em",
    color: "red"
  },
  inputWrapper: {
    flex: 1,
    maxWidth: "1000px",
    background: "white",
    borderRadius: "6px",
    padding: "4px 16px 8px",
    margin: "8px"
  },
  textField: {
    background: "white",
    borderRadius: "6px"
  }
}));

const NotePad = ({ setOpen, noteIdToBeEdited }) => {
  const classes = useStyles();
  const [text, setText] = useState("");
  const [nonprintable, setNonprintable] = useState(false);
  const [addNotes, isLoading] = useAddNotes();
  const [editNotes, editIsLoading] = useEditNotes();
  const dispatch = useDispatch();

  // Get note to be edited details from store
  const noteToBeEdited = useSelector(state =>
    noteIdToBeEdited ? state.notes.noteMap[noteIdToBeEdited] : null
  );

  // If there is a note to be edited load the text in the state variable
  useEffect(() => {
    if (noteToBeEdited) setText(noteToBeEdited.text);
  }, [noteToBeEdited]);

  // Handle input change.
  function handleChange(e) {
    const value = e.target.value;
    // If there are non printable characters set error in state. Remove error if new value meets condition.
    if (value && !/^[ -~\t\n\r]+$/.test(value)) {
      setNonprintable(true);
    } else if (nonprintable) {
      setNonprintable(false);
    }
    // Sets the input value to the text state variable.
    setText(e.target.value);
  }

  // Handle submit note.
  async function handleSubmit() {
    const trimmedText = text ? text.trim() : text;
    // If there is noteIdToBeEdited call edit function, otherwise call add note function
    if (noteIdToBeEdited) {
      await editNotes({ ...noteToBeEdited, text: trimmedText });
    } else {
      await addNotes({ text: trimmedText });
    }
    // Close note pad
    setOpen(false);
  }

  async function handleCancel() {
    // If there is a note to be edited remove from store
    if (noteIdToBeEdited) dispatch(setNoteToEdit(null));
    // Close note pad
    setOpen(false);
  }

  // Render note pad
  return (
    <>
      <div className={classes.inputWrapper}>
        <TextField
          id="note"
          className={classes.textField}
          label={noteIdToBeEdited ? "Edit Note" : "New Note"}
          value={text}
          multiline
          fullWidth
          error={nonprintable}
          margin="normal"
          onChange={handleChange}
        />
        {nonprintable && (
          <div className={classes.errorText}>
            Please remove non-printable characters
          </div>
        )}
      </div>
      <ActionButtons
        disableSubmit={nonprintable || !text || isLoading || editIsLoading}
        handleSubmit={handleSubmit}
        handleCancel={handleCancel}
      />
    </>
  );
};

NotePad.propTypes = {
  noteIdToBeEdited: PropTypes.string,
  setOpen: PropTypes.func
};

NotePad.defaultProps = {
  noteIdToBeEdited: "",
  setOpen: function() {}
};

export default memo(NotePad);
