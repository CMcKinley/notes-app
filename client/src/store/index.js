import { combineReducers, createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import notes from "./notes.js";

const store = combineReducers({ notes });

const composeEnhancers = composeWithDevTools({});

export default createStore(store, composeEnhancers(applyMiddleware(thunk)));
