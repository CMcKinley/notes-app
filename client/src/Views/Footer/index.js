import React, { memo, useState, useEffect } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Fab from "@material-ui/core/Fab";
import AddIcon from "@material-ui/icons/Add";
import { useSelector } from "react-redux";
import NotePad from "./NotePad";

const useStyles = makeStyles(() => ({
  container: {
    background: "#1976d2",
    display: "flex",
    justifyContent: "center"
  },
  appBar: {
    top: "auto",
    bottom: 0,
    display: "flex",
    justifyContent: "center",
    color: "#fff",
    backgroundColor: "#3f51b5",
    position: "fixed",
    width: "100%"
  },
  fab: {
    margin: "8px"
  }
}));

const Footer = () => {
  const classes = useStyles();
  const [open, setOpen] = useState(false);
  // Get the note Id that is to be edited from store
  const noteIdToBeEdited = useSelector(state => state.notes.noteToEdit);

  // If noteIdToBeEdited changes open the notepad. If the notepad was already open and noteIdToBeEdited is null, undefined or '' then close the notepad
  useEffect(() => {
    if (noteIdToBeEdited) {
      setOpen(true);
    } else if (!noteIdToBeEdited && open) {
      setOpen(false);
    }
  }, [noteIdToBeEdited]);

  // Render footer. Will display action button or notepad depending on the 'open' value
  return (
    <div className={classes.appBar}>
      {open ? (
        <NotePad setOpen={setOpen} noteIdToBeEdited={noteIdToBeEdited} />
      ) : (
        <Fab
          size="small"
          aria-label="add"
          onClick={() => setOpen(true)}
          className={classes.fab}
        >
          <AddIcon />
        </Fab>
      )}
    </div>
  );
};

export default memo(Footer);
