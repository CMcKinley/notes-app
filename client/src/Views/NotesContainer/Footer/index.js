import React, { memo, useState, useEffect } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Fab from "@material-ui/core/Fab";
import AddIcon from "@material-ui/icons/Add";
import { useSelector } from "react-redux";
import NotePad from "./NotePad";

const useStyles = makeStyles(() => ({
  container: {
    background: "#1976d2",
    display: "flex",
    justifyContent: "center"
  },
  appBar: {
    top: "auto",
    bottom: 0,
    display: "flex",
    justifyContent: "center",
    color: "#fff",
    backgroundColor: "#3f51b5",
    position: "fixed",
    width: "100%"
  },
  fab: {
    margin: "8px"
  }
}));

const Footer = () => {
  const classes = useStyles();
  const [open, setOpen] = useState(false);
  const noteIdToBeEdited = useSelector(state => state.notes.noteToEdit);

  useEffect(() => {
    if (noteIdToBeEdited) {
      setOpen(true);
    } else if (!noteIdToBeEdited && open) {
      setOpen(true);
    }
  }, [noteIdToBeEdited]);

  return (
    <div className={classes.appBar}>
      {open ? (
        <NotePad setOpen={setOpen} noteIdToBeEdited={noteIdToBeEdited} />
      ) : (
        <Fab
          size="small"
          aria-label="add"
          onClick={() => setOpen(true)}
          className={classes.fab}
        >
          <AddIcon />
        </Fab>
      )}
    </div>
  );
};

export default memo(Footer);
