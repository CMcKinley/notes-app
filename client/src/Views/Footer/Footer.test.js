import React from "react";
import renderer from "react-test-renderer";
import Footer from "./index.js";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { defultState } from "store/notes";
export const mockStore = configureStore([]);

const initialState = { notes: defultState };
const store = mockStore(initialState);

describe("Footer", () => {
  it("Renders correctly", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Footer />
      </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
