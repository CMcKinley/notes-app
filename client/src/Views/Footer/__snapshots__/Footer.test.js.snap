// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer Renders correctly 1`] = `
<div
  className="makeStyles-appBar-2"
>
  <button
    aria-label="add"
    className="MuiButtonBase-root MuiFab-root makeStyles-fab-3 MuiFab-sizeSmall"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </span>
  </button>
</div>
`;
